### PR TITLE
[experimental] PyTorch symmetric memory as an allocation provider

### PR DESCRIPTION
**Pull request by nirvedhmeshram**
## Summary

Adds `torch.distributed._symmetric_memory` as a second allocation provider, alongside the rocSHMEM one from #550. No changes to Iris kernels are needed.

## Motivation

#546 lists Torch Symmetric Memory and rocSHMEM as target providers of the same allocator-agnostic boundary. #550 did rocSHMEM; this is the other one, and it is the case that tests the boundary harder — rocSHMEM's mapping is uniform across its symmetric heap, so almost any table works. Torch's is not, so the per-allocation descriptor #546 specifies is load-bearing here rather than merely tidy.

As with #550, the whole integration is host-side. The test kernel calls unmodified `iris.store` on memory allocated entirely by torch.

## Technical details

`iris/experimental/torch_symmmem_provider.py` adds `TorchSymmMemProvider` with the same surface as the rocSHMEM provider — `allocate_symmetric`, `allocate_symmetric_map`, `symmetric_address_map` — matching `Iris.allocate_symmetric` from #549.

**The provider computes nothing.** Torch's rendezvous handle exposes `buffer_ptrs`: a per-peer list of pointers into this process's address space, whose local-rank entry already equals the tensor's own `data_ptr()`. That is precisely the table Iris translates against, so the provider reads it off the handle and asserts the invariant rather than deriving it. Where the rocSHMEM provider has to query `rocshmem_ptr` and cache per-peer offsets, this one has no pointer arithmetic at all.

The first commit moves `SymmetricAddressMap` out of `rocshmem_provider.py` into `iris/experimental/symmetric_memory.py`, which imports nothing but torch. Without that, using the descriptor would drag `rocshmem4py` into a provider that has no use for it. It is re-exported from `rocshmem_provider` so existing imports keep working.

### How it differs from the rocSHMEM provider

**Peer offsets are not shared between allocations.** Each torch allocation is a separate IPC mapping, not a window onto one linear heap, so a table only translates pointers into the allocation it describes. This is the opposite of rocSHMEM, where #550 documents and relies on tables being interchangeable.

This is not theoretical. An earlier revision of the test passed one allocation's table while storing to a different allocation — copied from the rocSHMEM test, where it is legitimate — and the kernel faulted.

The practical consequence is narrow: `iris.store` and `load` translate one pointer, so you pass the table of the allocation being addressed and a kernel can take several. `iris.copy` takes a single `heap_bases` and translates both `src` and `dst` against it, so under this provider it is only valid when both live in the same allocation.

**`symmetric_address_map` rejects foreign tensors.** Rendezvous is collective; deriving a handle on demand would hang whichever ranks did not ask. Handles are recorded at allocation time instead.

## Tests run in CI

Unlike #550, **no install and no image change are needed** — torch is already in the CI images, so the tests execute rather than skip with nothing added to `apptainer/iris.def` or `install_rocshmem.sh`.

That said, symmetric memory working on ROCm is not a given, and the module import is not a useful test of it: `torch.distributed._symmetric_memory` imports successfully on builds where allocation then fails. Measured on gfx950:

| torch | symmetric memory |
|---|---|
| `2.10.0+rocm7.2.1` — the Apptainer CI base | works |
| `2.13.0+rocm7.2` | works |
| `2.9.1+rocm7.1` — the Docker base | no; lacks `set_signal_pad_size`, and `rendezvous` fails with `HIP error: invalid argument` |
| `2.13.0+rocm7.15.0a` nightly | no; `hipErrorOutOfMemory` on a 4 KB allocation |

So the fixture probes by attempting a throwaway allocation and skips on failure. CI builds with Apptainer, so the tests **run**; on the Docker fallback they would skip cleanly rather than fail. As in #550, `importorskip` is inside the fixture rather than at module scope — a module-level one collects zero items, pytest returns exit code 5 (`NO_TESTS_COLLECTED`), and torchrun reports that as a child failure that fails the whole job.

One trap worth recording, since it cost a day: **do not call `symm_mem.set_backend()`**. The name `get_backend()` returns on ROCm is `'CUDA'` — the HIP IPC path — but `set_backend('CUDA')` is rejected with "SymmetricMemory does not find allocation backend CUDA", and forcing any other name selects a backend that then fails at allocation, which reads as "symmetric memory is unsupported on ROCm" when it is not. The provider never calls it.

## Test plan

```
python tests/run_tests_distributed.py tests/unittests/test_torch_symmmem_provider.py --num_ranks 2
```

Five tests: the table invariant, per-allocation anchoring, the reachability mask, rejection of foreign tensors, and unmodified `iris.store` over torch-allocated memory.

## Test results

On gfx950, on the same image CI uses (`rocm/pytorch:rocm7.2.1_ubuntu24.04_py3.14_pytorch_2.10.0`, torch `2.10.0+rocm7.2.1`):

| Ranks | Result |
|---|---|
| 2 | 5 passed |
| 4 | 5 passed |
| 8 | 5 passed |

All three rank counts were checked rather than just 2, because `symm_mem.rendezvous` is collective across the whole group, so rank count is a real variable and CI exercises 1, 2, 4 and 8.

## Not covered

- Intra-node only, set by the backend torch selects. The default reports as `'CUDA'` and is the HIP IPC path on ROCm; inter-node needs a transport this provider does not supply.
- `iris.copy` is not exercised. It is valid within a single allocation and invalid across two, per the mapping difference above — the restriction is documented but untested.
- Only `iris.store` is exercised. `load`, `put`, `get` and the atomics are single-translation and should behave identically.
- No views, no mixing two providers in one kernel, no imported tensors.

### CODE REVIEW

File: iris/experimental/torch_symm_mem_provider.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+                "translation would produce wild pointers."
+            )
+
+        return SymmetricAddressMap(
+            peer_bases=torch.tensor(bases, dtype=torch.int64, device=tensor.device),
+            local_rank=self.cur_rank,
+            allocation_base=tensor.data_ptr(),
+            allocation_bytes=tensor.numel() * tensor.element_size(),
+            direct=tuple(b != 0 for b in bases),
+        )
```
[the remark this answers is not in the document]
Comment (nirvedhmeshram) [reply iris/experimental/torch_symm_mem_provider.py:160]: Added the length check in ad31c4c8 — that part is a real hazard: Iris indexes this table by rank *inside the kernel*, so a short table reads past the end of the allocation rather than raising anything.

Not normalizing a `None` sentinel, though. Every build measured returns plain ints, one per rank (`buffer_ptrs=[126827509989376, 126827509956608]` at 2 ranks), so that would be handling for something unobserved. It would also be the wrong default if it ever did appear: mapping an unknown sentinel to `0` marks the peer *unreachable*, so a genuinely reachable peer would be silently skipped by callers checking `direct`. Raising on an unexpected type is the better failure. If a build turns up that does use a sentinel, that's the point to handle it — with a test.

File: tests/unittests/test_torch_symm_mem_provider.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+
+    if me == 0:
+        # results_bases, not data_bases: the store addresses `results`, and
+        # each allocation has its own mapping. tl.load of `data` is local and
+        # needs no translation.
+        _broadcast_kernel[(1,)](
+            data, results, results_bases, BLOCK_SIZE, me, num_ranks=ws, BLOCK_SIZE=BLOCK_SIZE, num_warps=4
+        )
+        torch.cuda.synchronize()
+    provider.barrier()
```
[the remark this answers is not in the document]
Comment (nirvedhmeshram) [reply tests/unittests/test_torch_symm_mem_provider.py:161]: Good catch — fixed in ad31c4c8.

The test synchronised on the writer before the barrier but never on the readers after it. `provider.barrier()` orders the ranks, but it doesn't drain the reading rank's stream, so the inbound writes weren't strictly ordered against work that rank queues next. Added `torch.cuda.synchronize()` on every rank after the barrier.

Worth noting it never failed in practice — 2, 4 and 8 ranks locally and in CI — so this is closing an ordering gap rather than fixing an observed flake.

File: iris/experimental/torch_symm_mem_provider.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        """
+        handle = self._handles.get(tensor.data_ptr())
+        if handle is None:
+            raise KeyError(
+                "tensor was not allocated by this provider, so its rendezvous "
+                "handle is unknown. Allocate through allocate_symmetric to have "
+                "the handle recorded; rendezvous cannot be repeated here because "
+                "it is collective."
+            )
+        return self._map_from_handle(tensor, handle)
```
[the remark this answers is not in the document]
Comment (nirvedhmeshram) [reply iris/experimental/torch_symm_mem_provider.py:129]: Leaving this uncached, same as the equivalent thread on #550.

The obvious cache key is `data_ptr()`, which aliases as soon as an allocation is freed and its address reused. A stale hit would hand back a table describing a *different* allocation — silently wrong pointers inside a kernel rather than an error. That's a bad trade for one `int64[num_ranks]` tensor: 64 bytes at 8 ranks.

On the cost itself: `allocate_symmetric` calls this once per allocation, so the normal path already builds it once. It isn't meant to be called per launch — the docstring says to hold the returned table for the allocation's lifetime rather than re-deriving it. A caller that does call it repeatedly is doing something the API doesn't ask for.

File: tests/unittests/test_torch_symmmem_provider.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        pytest.skip("needs torch.distributed; run via tests/run_tests_distributed.py")
+    if dist.get_world_size() < 2:
+        pytest.skip("needs at least 2 ranks (--num_ranks 2)")
+    if dist.get_backend() == "gloo":
+        pytest.skip("symmetric memory needs a device process group, not gloo")
+
+    # Imported here rather than at module scope so the tests are collected and
+    # individually skipped. A module-level importorskip collects zero items,
+    # which makes pytest exit 5 (NO_TESTS_COLLECTED) and fails the whole run.
+    from iris.experimental.torch_symmmem_provider import TorchSymmMemProvider
```
[the remark this answers is not in the document]
Comment (nirvedhmeshram) [reply tests/unittests/test_torch_symmmem_provider.py]: Agreed the triple m reads as a typo — renamed in ad31c4c8, though to `torch_symm_mem_provider` rather than `torch_symem_provider`.

torch abbreviates `torch.distributed._symmetric_memory` as `symm_mem` (double m, then `mem`), which is what the import line in this file already uses. `symem` would be a third spelling and further from torch's own. Test file and the README references renamed to match.

File: iris/experimental/torch_symm_mem_provider.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+        return tensor, self._map_from_handle(tensor, handle)
+
+    def symmetric_address_map(self, tensor: torch.Tensor) -> SymmetricAddressMap:
+        """Describe an already-allocated symmetric tensor.
+
+        The tensor must have been allocated through this provider, whose handle
+        is reused. Rendezvousing again here would be a collective call made from
+        whichever rank happened to ask.
+        """
+        handle = self._handles.get(tensor.data_ptr())
```
Comment (mawad-amd) [inline iris/experimental/torch_symm_mem_provider.py:121]: Do we want the providers to support tensors not allocated through the provider? I am thinking of a pattern like:

```
t = symm_mem.empty(*shape, dtype=dtype, device=self._device)
h = symm_mem.rendezvous(tensor, group=self._group)

provider.symmetric_address_map(t, h)
```
We can do a different PR later if needed. @drprajap let me know your thoughts too.

**Review by mawad-amd — APPROVED** [review]
Looks good. Thanks!

File: iris/experimental/torch_symm_mem_provider.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+rather than from Iris's own symmetric heap. No Iris device code changes are
+needed: iris.store, load and copy take ``heap_bases`` as a plain pointer argument
+and translate with
+
+    remote = peer_bases[to] + (ptr - peer_bases[local_rank])
+
+so any table satisfying ``peer_bases[local_rank] == local allocation base``
+drives them.
+
+Torch hands that table over directly. Rendezvous returns a handle whose
```
Comment (drprajap) [inline iris/experimental/torch_symm_mem_provider.py]: There is some duplication between here and README, can we only highlight important non obvious parts only? this seems excessively lengthy.

File: iris/experimental/torch_symm_mem_provider.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+``buffer_ptrs`` is a per-peer list of pointers into this process's address space,
+with the local rank's entry equal to the tensor's own ``data_ptr()`` -- exactly
+the invariant Iris needs. So this provider does no pointer arithmetic at all; it
+reads the table off the handle per allocation.
+
+That is worth preferring over the alternative even though the alternative also
+works. Peer offsets happen to be constant across allocations, so one table could
+be reused for all of them, but that only holds while every rank allocates in
+lockstep: ``symm_mem.empty`` is a local call and only ``rendezvous`` is
```
Comment (drprajap) [inline iris/experimental/torch_symm_mem_provider.py]: what is the difference here - README says both allocation and rendezvous are collective

File: iris/experimental/torch_symm_mem_provider.py [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+        self.num_ranks = dist.get_world_size(self._group)
+        self._device = (
+            torch.device(device) if device is not None else torch.device(f"cuda:{torch.cuda.current_device()}")
+        )
+        # data_ptr -> handle, so an allocation can be described again later
+        # without a second rendezvous. Rendezvous is collective: calling it from
+        # one rank alone would hang the others.
+        self._handles: dict[int, object] = {}
+
+    # ── table form ───────────────────────────────────────────────────────────
```
Comment (drprajap) [inline iris/experimental/torch_symm_mem_provider.py]: don't understand this, what is this and `descriptor form`

File: iris/experimental/torch_symm_mem_provider.py (hunk 5 above)
Comment (drprajap) [reply iris/experimental/torch_symm_mem_provider.py]: yes, we want to support already rendezvoused pytorch tensors, this would enhance support for usecase where we already own PyTorch symmetric tensor, which may have already called `symm_mem.empty()` and collectively `rendezvoused` it before Iris is involved. We should use exisitng peer pointers in that case instead of allocating second tensor and copying data.
okay  to follow-up with separate PR
